server/rooms.js

```diff
@@ -591,37 +591,55 @@ class GlobalRoom extends BasicRoom {
 		}
 		return this.formatList;
 	}
-	get configRankList() {
-		if (Config.nocustomgrouplist) return '';
-
-		// putting the resultant object in Config would enable this to be run again should config.js be reloaded.
-		if (Config.rankList) {
-			return Config.rankList;
-		}
-		let rankList = [];
-
-		for (let rank in Config.groups) {
-			if (!Config.groups[rank] || !rank) continue;
-
-			let tarGroup = Config.groups[rank];
-			let groupType = tarGroup.addhtml || (!tarGroup.mute && !tarGroup.root) ? 'normal' : (tarGroup.root || tarGroup.declare) ? 'leadership' : 'staff';
-
-			rankList.push({symbol: rank, name: (Config.groups[rank].name || null), type: groupType}); // send the first character in the rank, incase they put a string several characters long
+	/**
+	 * @param {AnyObject} group
+	 */
+	configGroupListMapper(group) {
+		let {symbol, id, name} = group;
+		let type;
+		if (group.editroom || group.makeroom || group.root) {
+			type = 'leadership';
+		} else if (!group.declare && (group.mute || group.ban)) {
+			type = 'staff';
+		} else if (group.punishgroup) {
+			type = 'punishment';
+		} else {
+			type = 'normal';
 		}
+		return {symbol, id, name, type};
+	}
+	get configGroupList() {
+		// Putting the resultant object in Config would enable this to be run again should config.js be reloaded.
+		if (Config.rankList) return Config.rankList;
 
 		const typeOrder = ['punishment', 'normal', 'staff', 'leadership'];
+		let rankList;
 
-		rankList = rankList.sort((a, b) => typeOrder.indexOf(b.type) - typeOrder.indexOf(a.type));
-
-		// add the punishment types at the very end.
-		for (let rank in Config.punishgroups) {
-			rankList.push({symbol: Config.punishgroups[rank].symbol, name: Config.punishgroups[rank].name, type: 'punishment'});
+		if (Config.grouplist) {
+			// @ts-ignore
+			rankList = Config.grouplist.map(this.configGroupListMapper).sort((a, b) => (
+				typeOrder.indexOf(b.type) - typeOrder.indexOf(a.type) ||
+					// @ts-ignore
+					Config.grouplist.findIndex(g => g.id === a.id) - Config.grouplist.findIndex(g => g.id === b.id)
+			));
+		} else {
+			const groups = Object.values(Config.groups);
+			const punishGroups = Object.values(Config.punishgroups);
+			rankList = [
+				...groups.map(this.configGroupListMapper),
```

Comment:
Seems like you could do the map on `rankList` instead of on both `groups` and `punishGroups` individually?

Reply:
Fixed now